lighthouse-core/gather/computed/dependency-graph/estimator/estimator.js

```diff
@@ -0,0 +1,291 @@
+      {
+        rtt: DEFAULT_RTT,
+        throughput: DEFAULT_THROUGHPUT,
+        fallbackTTFB: DEFAULT_FALLBACK_TTFB,
+        maximumConcurrentRequests: DEFAULT_MAXIMUM_CONCURRENT_REQUESTS,
+      },
+      options
+    );
+
+    this._rtt = this._options.rtt;
+    this._throughput = this._options.throughput;
+    this._fallbackTTFB = this._options.fallbackTTFB;
+    this._maximumConcurrentRequests = Math.min(
+      TcpConnection.maximumSaturatedConnections(this._rtt, this._throughput),
+      this._options.maximumConcurrentRequests
+    );
+  }
+
+  /**
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {number}
+   */
+  static getTTFB(record) {
+    const timing = record._timing;
+    return (timing && timing.receiveHeadersEnd - timing.sendEnd) || Infinity;
+  }
+
+  _initializeNetworkRecords() {
+    this._networkRecords = [];
+
+    this._graph.getRootNode().traverse(node => {
+      if (node.type === Node.TYPES.NETWORK) {
+        this._networkRecords.push(node.record);
+      }
+    });
+  }
+
+  _initializeNetworkConnections() {
+    const connections = new Map();
+    const recordsByConnection = groupBy(
+      this._networkRecords,
+      record => record.connectionId
+    );
+
+    for (const [connectionId, records] of recordsByConnection.entries()) {
+      const isTLS = TLS_SCHEMES.includes(records[0].parsedURL.scheme);
+
+      // We'll approximate how much time the server for a connection took to respond after receiving
+      // the request by computing the minimum TTFB time for requests on that connection.
+      //    TTFB = one way latency + server response time + one way latency
+      // Even though TTFB is greater than server response time, the RTT is underaccounted for by
+      // not varying per-server and so the difference roughly evens out.
+      // TODO(patrickhulce): investigate a way to identify per-server RTT
+      let estimatedResponseTime = records.reduce(
+        (min, record) => Math.min(min, Estimator.getTTFB(record)),
+        Infinity
+      );
+
+      // If we couldn't find a TTFB for the requests, use the fallback TTFB instead.
+      if (!Number.isFinite(estimatedResponseTime)) {
+        estimatedResponseTime = this._fallbackTTFB;
+      }
+
+      const connection = new TcpConnection(
+        this._rtt,
+        this._throughput,
+        estimatedResponseTime,
+        isTLS
+      );
+
+      connections.set(connectionId, connection);
+    }
+
+    this._connections = connections;
+    return connections;
+  }
+
+  _initializeAuxiliaryData() {
+    this._nodeTiming = new Map();
+    this._nodesCompleted = new Set();
+    this._nodesInProcess = new Set();
+    this._nodesInQueue = new Set(); // TODO: replace this with priority queue
+    this._connectionsInUse = new Set();
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  _enqueueNodeIfPossible(node) {
+    const dependencies = node.getDependencies();
+    if (
+      !this._nodesCompleted.has(node) &&
```

Comment:
if all dependencies must have been completed to be enqueued, shouldn't it not be possible for a node to already be completed here (since it's being added as a dependent of another node)

Reply:
This is cleaned up a little bit in #3162 but the summary here is that it was the only check against duplicate processing of a node. By definition, if a node has been completed, all of its dependencies have been completed, so while we're enqueing don't start on a node we've already finished (really should be if a node we've already started which is fixed later but it didn't have an impact when only network records were considered). 

Once again, this logic is re-arranged in the next PR to be clearer.

Reply:
> By definition, if a node has been completed, all of its dependencies have been completed, so while we're enqueing don't start on a node we've already finished

right, I was trying to say that always checking `dependencies.every(dependency => this._nodesCompleted.has(dependency))` before enqueuing should be sufficient, because it means if a dependent of a newly-completed node is being enqueued,  it can't have even been in the queue before (let alone completed), because one of its dependencies wasn't completed until just now.

We can wait as it changes, but it feels like it's a `// should never happen` situation and it should throw rather than ignoring just in case it does ever happen :)

Reply:
Oh oh I see my confusion here now :) 

In another version, this method was called in a loop over all nodes, not just the dependents of recently completed nodes. Yup, yup in this snapshot it's a 100% superfluous check.